src/transport/grpc/mod.rs

```diff
@@ -0,0 +1,304 @@
+        uuid: String,
+    ) -> Self {
+        Self {
+            shared: Arc::new(SharedGrpc {
+                client: Mutex::new(client),
+                stream: Mutex::new(stream),
+                uuid,
+            }),
+        }
+    }
+
+    /// Polls a message from the tonic stream and tries reattaching if necessary
+    ///
+    /// An [`Option`] is returned directly from the [`tonic::codec::Streaming::message`] method.
+    /// A result of [`Option::None`] signals a disconnectiond and should be handled by the caller
+    async fn next_message(&self) -> Result<Option<AstarteMessage>, GrpcTransportError> {
+        self.stream
+            .lock()
+            .await
+            .message()
+            .await
+            .map_err(GrpcTransportError::from)
+    }
+
+    async fn try_attach(
+        client: &mut MessageHubClient<tonic::transport::Channel>,
+        data: NodeData,
+    ) -> Result<tonic::codec::Streaming<AstarteMessage>, GrpcTransportError> {
+        client
+            .attach(tonic::Request::new(data.node))
+            .await
+            .map(|r| r.into_inner())
+            .map_err(GrpcTransportError::from)
+    }
+
+    async fn try_detach_attach(&self, data: NodeData) -> Result<(), GrpcTransportError> {
+        let mut client = self.client.lock().await;
+
+        // during the detach phase only the uuid is needed we can pass an empty array
+        // since the interfaces are alredy known to the message hub
+        // this api will change in the future
+        client
+            .detach(Node::new(self.uuid.clone(), &[] as &[Vec<u8>]))
+            .await
+            .map_err(GrpcTransportError::from)?;
+
+        *self.stream.lock().await = Grpc::try_attach(&mut client, data).await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Publish for Grpc {
+    async fn send_individual(&self, data: ValidatedIndividual<'_>) -> Result<(), crate::Error> {
+        self.client
+            .lock()
+            .await
+            .send(tonic::Request::new(data.try_into()?))
+            .await
+            .map(|_| ())
+            .map_err(|e| GrpcTransportError::from(e).into())
+    }
+
+    async fn send_object(&self, data: ValidatedObject<'_>) -> Result<(), crate::Error> {
+        self.client
+            .lock()
+            .await
+            .send(tonic::Request::new(data.try_into()?))
+            .await
+            .map(|_| ())
+            .map_err(|e| GrpcTransportError::from(e).into())
+    }
+}
+
+impl Deref for Grpc {
+    type Target = SharedGrpc;
+
+    fn deref(&self) -> &Self::Target {
+        &self.shared
+    }
+}
+
+#[async_trait]
+impl Receive for Grpc {
+    type Payload = GrpcReceivePayload;
+
+    async fn next_event<S>(
+        &self,
+        device: &SharedDevice<S>,
+    ) -> Result<ReceivedEvent<Self::Payload>, crate::Error>
+    where
+        S: PropertyStore,
+    {
+        loop {
+            match self.next_message().await? {
```

Comment:
The error case on `next_message` should also be handled without propagating it,  for example on device-runtime we handled the tonic status codes `Unavailable` and `Unknown` and in those specific cases we retried attaching the node to the message-hub. https://github.com/edgehog-device-manager/edgehog-device-runtime/blob/a31c8a6c91a08f382ab6411b77104b157ca1b371/src/data/astarte_message_hub_node.rs#L204

Reply:
I've changed the error handling code to take into account what you wrote, let me know if this works for you. It's still a bit different from the code you've linked but I think it makes sense to handle the `Ok(None)` case since the [documentation](https://docs.rs/tonic/latest/tonic/codec/struct.Streaming.html#method.message) declares that no more messages will be available.